cql3/restrictions/statement_restrictions.cc

```diff
@@ -1003,5 +1006,725 @@ ::shared_ptr<single_column_restriction> single_column_restriction::LIKE::apply_t
+bool limits(const binary_operator& opr, column_value_eval_bag bag) {
+    if (!opr.op->is_slice()) { // For EQ or NEQ, use equal().
+        throw std::logic_error("limits() called on non-slice op");
+    }
+    const auto& columns = std::get<0>(opr.lhs);
+    if (columns.size() > 1) {
+        auto multi = dynamic_pointer_cast<multi_item_terminal>(get_tuple(opr.rhs, bag.options));
+        if (!multi) {
+            throw exceptions::invalid_request_exception("multi-column comparison has right-hand side that isn't a tuple");
+        }
+        const auto& rhs = multi->get_elements();
+        if (rhs.size() != columns.size()) {
+            throw exceptions::invalid_request_exception(
+                    format("tuple comparison size mismatch: {} elements on left-hand side, {} on right",
+                           columns.size(), rhs.size()));
+        }
+        for (size_t i = 0; i < rhs.size(); ++i) {
+            const auto cmp = get_value_comparator(columns[i])->as_tri_comparator()(
+                    // CQL dictates that columns[i] is a clustering column and non-null.
+                    *get_value(columns[i], bag),
+                    *rhs[i]);
+            // If the components aren't equal, then we just learned the LHS/RHS order.
+            if (cmp < 0) {
+                if (*opr.op == operator_type::LT || *opr.op == operator_type::LTE) {
+                    return true;
+                } else if (*opr.op == operator_type::GT || *opr.op == operator_type::GTE) {
+                    return false;
+                } else {
+                    throw std::logic_error("Unknown slice operator");
+                }
+            } else if (cmp > 0) {
+                if (*opr.op == operator_type::LT || *opr.op == operator_type::LTE) {
+                    return false;
+                } else if (*opr.op == operator_type::GT || *opr.op == operator_type::GTE) {
+                    return true;
+                } else {
+                    throw std::logic_error("Unknown slice operator");
+                }
+            }
+            // Otherwise, we don't know the LHS/RHS order, so check the next component.
+        }
+        // Getting here means LHS == RHS.
+        return *opr.op == operator_type::LTE || *opr.op == operator_type::GTE;
+    } else if (columns.size() == 1) {
+        auto lhs = get_value(columns[0], bag);
+        if (!lhs) {
+            lhs = bytes(); // Compatible with old code, which feeds null to type comparators.
+        }
+        auto tup = dynamic_pointer_cast<tuples::value>(get_tuple(opr.rhs, bag.options));
+        // Assume this is an external query WHERE (ck1)>(123), rather than an internal query WHERE col>(123),
+        // because internal queries have no reason to use single-element tuples.
+        //
+        // TODO: make the two cases distinguishable.
+        auto rhs = (tup && tup->size() == 1) ? tup->get_elements()[0]
+                : to_bytes_opt(opr.rhs->bind_and_get(bag.options));
```

Comment:
> I think (not sure) I get where this "if possible" comes from now. But maybe we should invert the order and first try the non-tuple case.

How would we know it's the non-tuple case, other than checking the RHS for tuple non-terminals?

> does the grammar (and glue) erase the fact that ck1 was enclosed in parentheses?

For now, yes (sadly).  I want to change that in the future, as explained in another comment upthread.

> Where does that happen?

It's dictated by the definition of binary_operator LHS and RHS.  The LHS has a vector of column_values, assuming a single-element vector decays into a plain reference.  Similarly, the RHS is (dynamically typed) `term`.  I already have plans to evolve the RHS into statically typed; do you think LHS should also change to distinguish between single-element tuples and plain column names?

Reply:
> > I think (not sure) I get where this "if possible" comes from now. But maybe we should invert the order and first try the non-tuple case.
> 
> How would we know it's the non-tuple case, other than checking the RHS for tuple non-terminals?

If the grammar/preparer prevents non-tuples, then they can't happen.

There are three cases:

     WHERE tup = (1, 2, 3)

clearly this is a tuple and we can reject non-tuples.

     WHERE tup = (1, :foo, :bar)

same

    WHERE tup = :baz
 
Here, we assign a tuple type to the bind variable, so it can't be a non-tuple (we'll interpret whatever the user sends according to the tuple serialization rules).


> 
> > does the grammar (and glue) erase the fact that ck1 was enclosed in parentheses?
> 
> For now, yes (sadly). I want to change that in the future, as explained in another comment upthread.

Ok. Please add detailed code comments.

> 
> > Where does that happen?
> 
> It's dictated by the definition of binary_operator LHS and RHS. The LHS has a vector of column_values, assuming a single-element vector decays into a plain reference. Similarly, the RHS is (dynamically typed) `term`. I already have plans to evolve the RHS into statically typed;

Note that by itself this isn't enough. If you have binary_operator(eq, variant<scalar, tuple>. variant<scalar_term, tuple_term>) then it explodes into four different combinations, two of which are legal. You need scalar_eq and tuple_eq in order to prevent the illegal combinations.

> do you think LHS should also change to distinguish between single-element tuples and plain column names?

First, there is ambiguity between parentheses for grouping and parentheses for tuples. Most languages use (x,) for single-element tuples to disambiguate.

Something like

    WHERE (x, y) > (:foo, :bar)

should be expressed as

    binary_operation(
          greater_than,
          make_tuple(
                column(x),
                column(y))
          ),
          make_tuple(
                marker(:foo),
                marker(:bar)
          ),
    )

(the second make_tuple is now a delayed_value, and the first is likey destroyed early by a check that the column names tuple consists of a prefix of the clustering keys)

Does the current code remember the type of the bind variable as a tuple or as the column type?

Reply:
> There are three cases:

Only if you don't count IN, which aggressively multiplies that number.

Reply:
> If you have binary_operator(eq, variant<scalar, tuple>. variant<scalar_term, tuple_term>) then it explodes into four different combinations, two of which are legal.

Three?  I thought internal queries can create binary_operator(eq, column(a), make_tuple(1,2,3))

> Does the current code remember the type of the bind variable as a tuple or as the column type?

As a tuple (see cql3::tuples inner classes that inherit from non_terminal).  The column type is accessed via column_definition, which is stored separately.

Reply:
For the record, here is how the RHS term hierarchy currently works:

|CQL|`term` produced|how to get bytes_opt from it|
|---|---|---|
|`a>=1`|constant|bind_and_get|
|`a>=?`|constant::marker|bind_and_get|
|`a IN (1,2)`|lists::delayed_value|get_elements returns `vector<term>`, bind_and_get each element|
|`a IN (1,2,?)`|same as above|same as above|
|`a IN ?`|lists::marker| bind() returns lists::value, call get_elements on it|
|`(a,b) >= (1,2)`|tuples::value|get_elements|
|`(a,b) >= (1,2,?)`|tuples::delayed_value|bind returns tuples::value, then see above|
|`(a,b) >= ?`|tuples::marker|same as above|
|`(a,b) IN ((1,1), (2,?), ?)`|lists::delayed_value|get_elements returns `vector<term>`, each is tuples::value/delayed_value/marker (respectively), see above|
|`(a,b) IN ?`|tuples::in_marker|bind returns tuples::in_value, whose get_split_values returns `vector<vector<bytes_opt>>`|

Each `constant` above becomes set/list/map, when the column is of collection type.

Reply:
> For the record, here is how the RHS term hierarchy currently works:
> 

Thanks, this is helpful. Some randome comments:

> CQL	`term` produced	how to get bytes_opt from it
> `a>=1`	constant	bind_and_get
> `a>=?`	constant::marker	bind_and_get
> `a IN (1,2)`	lists::delayed_value	get_elements returns `vector<term>`, bind_and_get each element

Seems like this should return a constant (during prepare time we can notice there are no markers).

> `a IN (1,2,?)`	same as above	same as above

Here delayed_value is justified.

Note I don't see delayed_value::get_elements(), just value::get_elements().

> `a IN ?`	lists::marker	bind() returns lists::value, whose get_elements returns `vector<bytes_opt>`
> `(a,b) >= (1,2)`	tuples::value	get_elements returns `vector<bytes_opt>`
> `(a,b) >= (1,2,?)`	tuples::delayed_value	bind returns tuples::value, then see above
> `(a,b) >= ?`	tuples::marker	same as above
> `(a,b) IN ((1,1), (2,?), ?)`	lists::delayed_value	get_elements returns `vector<term>`, each is tuples::value/delayed_value/marker (respectively), see above
> `(a,b) IN ?`	tuples::in_marker	bind returns tuples::in_value, whose get_split_values returns `vector<vector<bytes_opt>>`
> Each `constant` above becomes set/list/map, when the column is of collection type.

Everything here looks consistent with static typing.

Reply:
> Note I don't see delayed_value::get_elements(), just value::get_elements().

I added that in this PR: [lists.hh:109](https://github.com/dekimir/scylla/commit/4b4432873f9f013344cd20a3f08460d3d35a5561#diff-ee1c8dfbe7135ec9ac3c23a44c88b64eR109).

Reply:
Anticipating your next question: before my change, it worked by having a `vector<term>` member in the [fat class](https://sourcegraph.com/github.com/scylladb/scylla@ecc83e83e586c10271f8e08946103cfc10995d62/-/blob/cql3/single_column_relation.cc#L95).  Now we only have a single term on RHS.

Reply:
I see, thanks. Without this new delayed_value::get_elements(), you would have had to bind_and_get(), static_cast to lists::value, and get_elements() from there, yes?

(the new delayed_value::get_elements() is fine, asking to validate my understanding; in fact it looks more consistent with static knowledge of the types)

Reply:
> I see, thanks. Without this new delayed_value::get_elements(), you would have had to bind_and_get(), static_cast to lists::value, and get_elements() from there, yes?

Yes, exactly.

Reply:
Minimized and explained dynamic casts.